climakitae/explore/typical_meteorological_year.py

```diff
@@ -0,0 +1,804 @@
+
+    # Turn the CDF array into xarray DataArray
+    # New dimension is the bin values
+    cdf_da = xr.DataArray(
+        [bins_count[1:], cdf_np],
+        dims=["data", "bin_number"],
+        coords={
+            "data": ["bins", "probability"],
+        },
+    )
+    cdf_da.name = da.name
+    return cdf_da
+
+
+def _get_cdf_by_sim(da: xr.DataArray) -> xr.DataArray:
+    """Function to help get_cdf.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+        Hourly data
+
+    Returns
+    -------
+    xr.DataArray
+    """
+    # Group the DataArray by simulation
+    return da.groupby("simulation").map(_compute_cdf)
+
+
+def _get_cdf_by_mon_and_sim(da: xr.DataArray) -> xr.DataArray:
+    """Function to help get_cdf.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+        Hourly data
+
+    Returns
+    -------
+    xr.DataArray
+    """
+    # Group the DataArray by month in the year
+    return da.groupby("time.month").map(_get_cdf_by_sim)
+
+
+def get_cdf(ds: xr.DataArray) -> xr.Dataset:
+    """Get the cumulative density function.
+
+    Parameters
+    -----------
+    ds: xr.DataArray
+        Input data for which to compute CDF
+
+    Returns
+    -------
+    xr.Dataset
+    """
+    return ds.map(_get_cdf_by_mon_and_sim)
+
+
+def get_cdf_monthly(ds: xr.DataArray) -> xr.Dataset:
+    """Get the cumulative density function by unique mon-yr combos
+
+    Parameters
+    -----------
+    ds: xr.DataArray
+        Input data for which to compute CDF
+
+    Returns
+    -------
+    xr.Dataset
+    """
+
+    def get_cdf_mon_yr(da):
+        return da.groupby("time.year").map(_get_cdf_by_mon_and_sim)
+
+    return ds.map(get_cdf_mon_yr)
+
+
+def remove_pinatubo_years(ds: xr.Dataset) -> xr.Dataset:
+    """Drop years after Pinatubo eruption from dataset.
+
+    Parameters
+    ----------
+    ds: xr.Dataset
+
+    Returns
+    -------
+    xr.Dataset
+    """
+    ds = ds.where((~ds.year.isin([1991, 1992, 1993, 1994])), np.nan, drop=True)
+    return ds
+
+
+def fs_statistic(cdf_climatology: xr.Dataset, cdf_monthly: xr.DataArray) -> xr.Dataset:
+    """Calculates the Finkelstein-Schafer statistic.
+
+    Absolute difference between long-term climatology and candidate CDF, divided by number of days in month.
+
+    Parameters
+    -----------
+    cdf_climatology: xr.Dataset
+       Climatological CDF dataset (get_cdf result)
+    cdf_monthly: xr.Dataset
+       Monthly CDF dataset (get_cdf_monthly result)
+
+    Returns
+    -------
+    xr.Dataset
+        F-S statistic
+    """
+    days_per_mon = xr.DataArray(
+        data=[31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31],
+        coords={"month": np.arange(1, 13)},
+    )
+    fs_stat = abs(cdf_monthly - cdf_climatology).sel(data="probability") / days_per_mon
+    return fs_stat
+
+
+def compute_weighted_fs(da_fs: xr.Dataset) -> xr.Dataset:
+    """Weights the Finkelstein-Schafer (F-S) statistic based on TMY3 methodology.
```

Comment:
I'd throw in a reference into the docstring about these being the recommended / actual NREL weights for official TMY files: https://nsrdb.nrel.gov/data-sets/tmy